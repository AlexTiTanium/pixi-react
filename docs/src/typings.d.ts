declare module '*.md';
declare module '!!raw-loader!.*';
