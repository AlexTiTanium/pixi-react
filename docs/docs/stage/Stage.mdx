import { EmbeddedEditor } from '@site/src/components/Editor/EmbeddedEditor';
import StageFile from '!!raw-loader!./Stage';
import RAFFile from '!!raw-loader!./raf';
import RAF2File from '!!raw-loader!./raf2';

# Stage

The Stage component renders a `PIXI.Application` onto a canvas element and makes the `PIXI.Application` available for child components using the React's Context API.

## Props

| Prop                      | Description                                                                                | Default value |
| ------------------------- | ------------------------------------------------------------------------------------------ | ------------- |
| `width`                   | the width of the renderers view                                                            | `800`         |
| `height`                  | the height of the renderers view                                                           | `600`         |
| `onMount`                 | callback function for the created app instance                                             |               |
| `onUnMount`               | callback function when the Stage gets unmounted                                            |               |
| `raf`                     | use the internal PIXI ticker (requestAnimationFrame)                                       | `true`        |
| `renderOnComponentChange` | render stage on component changes. Only useful in combination with `raf` disabled          | `true`        |
| `options`                 | see [PIXI.Application options](https://pixijs.download/release/docs/PIXI.Application.html) |               |

## Usage

```jsx
import { render } from 'react-dom'
import { Stage, Container, Sprite } from '@pixi/react'

const App = () => (
  <Stage width={500} height={400}>
    { // Pixi React components here... }
  </Stage>
)

render(<App />, document.getElementById('root'))
```

## Example

<EmbeddedEditor
    height={350}
    files={{ 'App.js': StageFile }}
    dependencies={{
        '@pixi/react-animated': '^7',
        'react-spring': '^9',
        'pixi.js': '^7',
        '@pixi/react': '^7',
        lodash: '^4',
    }}
/>
## Custom Updates

By default the `<Stage>` component creates a `Pixi.Application` and updates the stage in a `requestAnimationFrame` loop.
It tries to update and render the stage at the max browser rate (~60fps).
It updates and renders the stage, even when there are no visual changes.
This might result in faster running cooling fans on your computer due the overhead of GPU and CPU calculations.

### Update stage on React component changes

If you want to know when components have changed, you can listen to the `__REACT_PIXI_REQUEST_RENDER__` event on the root container (app.stage).

> Disable `raf` and enable `renderOnComponentChange`:

<EmbeddedEditor height={350} files={{ 'App.js': RAFFile }} />

### Update stage manually

> Disable `raf` and `renderOnComponentChange`:

<EmbeddedEditor height={400} files={{ 'App.js': RAF2File }} />

## Accessing Application

To access the `Application` in your child components you can use one of the following:

#### AppConsumer

```jsx
import { AppConsumer } from '@pixi/react';

const MyComponent = () => <AppConsumer>{(app) => <OtherComponent app={app} />}</AppConsumer>;
```

#### withPixiApp

```jsx
import { withPixiApp } from '@pixi/react';

const MyComponent = withPixiApp(OtherComponent);
```

#### useApp

```jsx
import { useApp } from '@pixi/react';

const MyComponent = () => {
    const app = useApp();
    // app => PIXI.Application
};
```
