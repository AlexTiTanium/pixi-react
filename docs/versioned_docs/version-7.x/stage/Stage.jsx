import { Stage, Sprite } from '@pixi/react';
import lodash from 'lodash';

const count = 10;
const width = 300;
const height = 300;
const stageProps = {
    height,
    width,
    options: {
        backgroundAlpha: 0,
        antialias: true,
    },
};

export default function StageExample() {
    return (
        <Stage {...stageProps}>
            {lodash.times(count, (i) => (
                <Sprite
                    key={i}
                    image='https://pixijs.io/pixi-react/img/coin.png'
                    scale={(360 / count) * 0.004}
                    anchor={0.5}
                    rotation={i * (360 / count) * (Math.PI / 180)}
                    x={width / 2 + Math.cos(i * (360 / count) * (Math.PI / 180)) * 100}
                    y={height / 2 + Math.sin(i * (360 / count) * (Math.PI / 180)) * 100}
                />
            ))}
        </Stage>
    );
}
