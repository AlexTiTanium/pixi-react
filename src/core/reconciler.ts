import packageData from '../../package.json' with { type: 'json' };
import { afterActiveInstanceBlur } from '../helpers/afterActiveInstanceBlur';
import { appendChild } from '../helpers/appendChild';
import { beforeActiveInstanceBlur } from '../helpers/beforeActiveInstanceBlur';
import { clearContainer } from '../helpers/clearContainer';
import { commitUpdate } from '../helpers/commitUpdate';
import { createInstance } from '../helpers/createInstance';
import { createReconciler } from '../helpers/createReconciler';
import { createTextInstance } from '../helpers/createTextInstance';
import { detachDeletedInstance } from '../helpers/detachDeletedInstance';
import { finalizeInitialChildren } from '../helpers/finalizeInitialChildren';
import { getChildHostContext } from '../helpers/getChildHostContext';
import { getCurrentUpdatePriority } from '../helpers/getCurrentUpdatePriority';
import { getInstanceFromNode } from '../helpers/getInstanceFromNode';
import { getInstanceFromScope } from '../helpers/getInstanceFromScope';
import { getPublicInstance } from '../helpers/getPublicInstance';
import { getRootHostContext } from '../helpers/getRootHostContext';
import { hideInstance } from '../helpers/hideInstance';
import { hideTextInstance } from '../helpers/hideTextInstance';
import { insertBefore } from '../helpers/insertBefore';
import { maySuspendCommit } from '../helpers/maySuspendCommit';
import { preloadInstance } from '../helpers/preloadInstance';
import { prepareForCommit } from '../helpers/prepareForCommit';
import { preparePortalMount } from '../helpers/preparePortalMount';
import { prepareScopeUpdate } from '../helpers/prepareScopeUpdate';
import { removeChild } from '../helpers/removeChild';
import { requestPostPaintCallback } from '../helpers/requestPostPaintCallback';
import { resetAfterCommit } from '../helpers/resetAfterCommit';
import { resetFormInstance } from '../helpers/resetFormInstance';
import { resolveEventTimeStamp } from '../helpers/resolveEventTimeStamp';
import { resolveEventType } from '../helpers/resolveEventType';
import { resolveUpdatePriority } from '../helpers/resolveUpdatePriority';
import { setCurrentUpdatePriority } from '../helpers/setCurrentUpdatePriority';
import { shouldAttemptEagerTransition } from '../helpers/shouldAttemptEagerTransition';
import { shouldSetTextContent } from '../helpers/shouldSetTextContent';
import { startSuspendingCommit } from '../helpers/startSuspendingCommit';
import { suspendInstance } from '../helpers/suspendInstance';
import { trackSchedulerEvent } from '../helpers/trackSchedulerEvent';
import { unhideInstance } from '../helpers/unhideInstance';
import { unhideTextInstance } from '../helpers/unhideTextInstance';
import { waitForCommitToBeReady } from '../helpers/waitForCommitToBeReady';
import { type HostConfig } from '../typedefs/HostConfig';

const reconcilerConfig = {
    isPrimaryRenderer: false,
    noTimeout: -1,
    NotPendingTransition: null,
    supportsHydration: false,
    supportsMutation: true,
    supportsPersistence: false,
    warnsIfNotActing: false,

    afterActiveInstanceBlur,
    appendChild,
    appendChildToContainer: appendChild,
    appendInitialChild: appendChild,
    beforeActiveInstanceBlur,
    cancelTimeout: clearTimeout,
    clearContainer,
    commitUpdate,
    createInstance,
    createTextInstance,
    hideTextInstance,
    unhideTextInstance,
    detachDeletedInstance,
    finalizeInitialChildren,
    getChildHostContext,
    getCurrentUpdatePriority,
    getInstanceFromNode,
    getInstanceFromScope,
    getPublicInstance,
    getRootHostContext,
    hideInstance,
    insertBefore,
    insertInContainerBefore: insertBefore,
    maySuspendCommit,
    preloadInstance,
    prepareForCommit,
    preparePortalMount,
    prepareScopeUpdate,
    removeChild,
    removeChildFromContainer: removeChild,
    requestPostPaintCallback,
    resetAfterCommit,
    resetFormInstance,
    resolveEventTimeStamp,
    resolveEventType,
    resolveUpdatePriority,
    scheduleTimeout: setTimeout,
    shouldAttemptEagerTransition,
    setCurrentUpdatePriority,
    shouldSetTextContent,
    startSuspendingCommit,
    suspendInstance,
    trackSchedulerEvent,
    unhideInstance,
    waitForCommitToBeReady,
};

const reconciler = createReconciler<
    HostConfig['type'],
    HostConfig['props'],
    HostConfig['containerInstance'],
    HostConfig['instance'],
    HostConfig['textInstance'],
    HostConfig['suspenseInstance'],
    HostConfig['hydratableInstance'],
    HostConfig['formInstance'],
    HostConfig['publicInstance'],
    HostConfig['hostContext'],
    HostConfig['childSet'],
    HostConfig['timeoutHandle'],
    HostConfig['noTimeout'],
    HostConfig['transitionStatus']
>(reconcilerConfig);

reconciler.injectIntoDevTools({
    bundleType: process.env.NODE_ENV === 'production' ? 0 : 1,
    rendererPackageName: '@pixi/react',
    version: packageData.version,
});

export { reconciler };
