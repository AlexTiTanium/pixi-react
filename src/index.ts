export * from './reconciler';
export * from './render';
