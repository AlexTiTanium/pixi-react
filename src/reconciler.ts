// Module imports
import * as PIXI from 'pixi.js';
import Reconciler from 'react-reconciler';
import { DefaultEventPriority } from 'react-reconciler/constants';
// Local imports
import { appendChild } from './helpers/appendChild';
import { applyProps } from './helpers/applyProps';
import { convertStringToPascalCase } from './helpers/convertStringToPascalCase';
import { removeChild } from './helpers/removeChild';

import type { HostConfig } from 'react-reconciler';

export type Props = Record<string, any>;

/** @type {Reconciler.HostConfig} */
const reconcilerConfig: HostConfig<
string, // Type
Props, // Props
PIXI.Container, // Container
PIXI.Container, // Instance
any, // TextInstance
unknown, // SuspenseInstance
unknown, // HydratableInstance
unknown, // PublicInstance
any, // HostContext
any, // UpdatePayload
any, // ChildSet
number, // TimeoutHandle
number // NoTimeout
> = {
    isPrimaryRenderer: false,
    noTimeout: -1,
    supportsHydration: false,
    supportsMutation: true,
    supportsPersistence: false,

    appendChild,
    appendChildToContainer: appendChild,
    appendInitialChild: appendChild,
    cancelTimeout: clearTimeout,
    removeChild,
    removeChildFromContainer: removeChild,
    scheduleTimeout: setTimeout,

    afterActiveInstanceBlur() { /** */ },
    beforeActiveInstanceBlur() { /** */ },
    createTextInstance() { /** */ },
    detachDeletedInstance() { /** */ },
    getInstanceFromScope() { return null; },
    insertBefore() { /** */ },
    preparePortalMount() { /** */ },
    prepareScopeUpdate() { /** */ },

    clearContainer()
    {
        return false;
    },
    finalizeInitialChildren()
    {
        return false;
    },
    getChildHostContext()
    {
        return {};
    },
    getCurrentEventPriority()
    {
        return DefaultEventPriority;
    },
    getInstanceFromNode(): null
    {
        return null;
    },
    getPublicInstance(instance: PIXI.Container)
    {
        return instance;
    },
    getRootHostContext()
    {
        return {};
    },
    prepareForCommit()
    {
        return {};
    },
    prepareUpdate()
    {
        return {};
    },
    resetAfterCommit()
    {
        return {};
    },
    shouldSetTextContent()
    {
        return false;
    },

    /**
	 *
	 * @param {*} instance
	 * @param {*} _updatePayload Unused.
	 * @param {*} _type Unused.
	 * @param {*} oldProps
	 * @param {*} newProps
	 */
    commitUpdate(instance, _updatePayload, _type, oldProps, newProps)
    {
        // This is where we mutate Pixi.js objects in the render phase
        instance.busy = true;
        applyProps(instance, newProps, oldProps);
        instance.busy = false;
    },

    /**
	 *
	 * @param {*} type
	 * @param {*} props
	 * @returns
	 */
    createInstance(type: string, props)
    {
        const {
            object,
            args,
            // ...remainingProps
        } = props;

        // Convert lowercase primitive to PascalCase
        const name = convertStringToPascalCase(type);

        // Get class from Pixi.js namespace
        const Target = PIXI[name as 'Container'];

        // Validate Pixi.js elements
        if (type !== 'primitive' && !Target)
        {
            throw Error(`${type} is not a part of the Pixi.js namespace.`);
        }

        // Validate primitives
        if (type === 'primitive' && !object)
        {
            throw Error('"object" must be set when using primitives.');
        }

        // Create instance
        const instance: PIXI.Container = object || (Array.isArray(args) ? new Target(...args) : new Target(args));

        // Set initial props
        applyProps(instance, props, {});

        return instance;
    },
};

export const reconciler = Reconciler(reconcilerConfig);
