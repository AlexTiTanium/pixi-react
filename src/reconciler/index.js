import Reconciler from 'react-reconciler'
import pkg from '../../package.json'
import hostconfig from './hostconfig'

export const PixiFiber = Reconciler(hostconfig)
export const VERSION = pkg.version
export const REACT_DOM_VERSION = pkg.devDependencies['react-dom'].replace(/[^0-9.]/g, '')
export const PACKAGE_NAME = pkg.name
