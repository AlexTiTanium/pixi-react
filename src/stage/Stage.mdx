---
name: 'Stage'
route: '/stage'
---

import { Playground } from 'docz'
import times from 'lodash.times'
import { Stage, Sprite } from '../../docz-rp'

# Stage

The Stage component renders a `PIXI.Application` onto a canvas element and makes the `PIXI.Application` available for child components using the React's Context API.

## Props

| Prop  | Description | Default value |
| ------------- | ------------- | ------------- |
| `width`  | the width of the renderers view  | `800` |
| `height`  | the height of the renderers view  | `600` |
| `onMount`  | callback function for the created app instance  | |
| `onUnMount`  | callback function when the Stage gets unmounted  | |
| `raf`  | use the internal PIXI ticker (requestAnimationFrame)  | `true` |
| `renderOnComponentChange` |  render stage on Stage changes? only useful in combination with `raf` | `true` |
| `options` | see [PIXI.Application options](http://pixijs.download/release/docs/PIXI.Application.html) | |

## Usage

```jsx
import { render } from 'react-dom'
import { Stage, Container, Sprite } from '@inlet/react-pixi'

const App = () => (
  <Stage width={500} height={400}>
    { // React-Pixi components here... }
  </Stage>
)

render(<App />, document.getElementById('root'))
```

## Example

<Playground>
  {() => {
    const count = 10
    const width = 300
    const height = 300
    const stageProps = {
      height,
      width,
      options: {
        transparent: true,
        antialias: true
      }
    }
    return (
      <Stage
        {...stageProps}
      >
        {
          times(count, i => (
            <Sprite key={i}
                    image="https://s3-us-west-2.amazonaws.com/s.cdpn.io/693612/coin.png"
                    scale={(360/count) * 0.004}
                    anchor={0.5}
                    rotation={(i * (360/count)) * (Math.PI / 180)}
                    x={width/2 + Math.cos((i * (360/count)) * (Math.PI / 180)) * 100}
                    y={height/2 + Math.sin((i * (360/count)) * (Math.PI / 180)) * 100}  />
          ))
        }
      </Stage>
    )
  }}
</Playground>

