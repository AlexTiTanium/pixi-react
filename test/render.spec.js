import * as PIXI from 'pixi.js'
import React from 'react'
import Stage from '../src/stage'
import { AppConsumer, AppProvider, Container, Text, withPixiApp } from '../src'
import { render, roots } from '../src/render'
import { PACKAGE_NAME, PixiFiber, REACT_DOM_VERSION } from '../src/reconciler'

const app = new PIXI.Application()
const callback = jest.fn()
const element = () => (
  <Stage>
    <Text text="Hello Word!" />
  </Stage>
)

jest.mock('../src/reconciler')

describe('render', () => {
  let spies

  beforeEach(() => {
    roots.clear()
    jest.resetAllMocks()

    spies = {
      createContainer: jest.fn(),
      updateContainer: jest.fn(),
      getPublicInstance: jest.fn(),
      injectIntoDevTools: jest.fn(),
    }

    PixiFiber.mockImplementation(() => ({ ...jest.requireActual('../src/reconciler').PixiFiber(), ...spies }))

    spies.createContainer.mockReturnValue({ current: { child: { tag: 'TEXT' } } })
    spies.updateContainer.mockImplementation((element, root, _, c) => c())
  })

  test('invariant container', () => {
    expect(() => render('something', null)).toThrow(
      'Invalid argument `container`, expected instance of `PIXI.Container`'
    )
  })

  test('call createContainer', () => {
    render(element, app.stage, callback)
    expect(spies.createContainer).toHaveBeenCalledTimes(1)
    expect(spies.createContainer).toHaveBeenLastCalledWith(app.stage)
  })

  test('call updateContainer', () => {
    render(element, app.stage, callback)
    const root = roots.values().next().value

    expect(spies.updateContainer).toHaveBeenCalledTimes(1)
    expect(spies.updateContainer).toHaveBeenLastCalledWith(element, root, undefined, callback)
  })

  test('call injectDevtools', () => {
    render(element, app.stage, callback)

    expect(spies.injectIntoDevTools).toHaveBeenCalledTimes(1)
    expect(spies.injectIntoDevTools.mock.calls).toMatchSnapshot()
  })

  test('invoke callback in updateContainer', () => {
    render(element, app.stage, callback)
    expect(callback).toHaveBeenCalledTimes(1)
  })

  test('store root', () => {
    render(element, app.stage, callback)
    expect(roots.values().next().value).toEqual({ current: { child: { tag: 'TEXT' } } })
  })

  test('does not create root if it is already present', () => {
    const root = { current: { child: { tag: 'CUSTOM' } } }
    roots.set(app.stage, root)

    render(element, app.stage, callback)
    expect(spies.createContainer).toHaveBeenCalledTimes(0)
  })

  describe('passdown `PIXI.Application`', () => {
    beforeEach(() => {
      const pf = jest.requireActual('../src/reconciler').PixiFiber()
      spies.createContainer.mockImplementation(pf.createContainer)
      spies.updateContainer.mockImplementation(pf.updateContainer)
    })

    test('via `AppConsumer`', () => {
      const fn = jest.fn(() => <Text text="hi" />)

      render(
        <AppProvider value={app}>
          <Container>
            <AppConsumer>{fn}</AppConsumer>
          </Container>
        </AppProvider>,
        app.stage,
        callback
      )

      expect(fn).toHaveBeenCalledTimes(1)
      expect(fn).toHaveBeenCalledWith(app)
    })

    test('via `withPixiApp`', () => {
      const fn = jest.fn(() => <Text text="hi" />)
      const Comp = withPixiApp(({ app }) => fn(app))

      render(
        <AppProvider value={app}>
          <Container>
            <Comp />
          </Container>
        </AppProvider>,
        app.stage,
        callback
      )

      expect(fn).toHaveBeenCalledTimes(1)
      expect(fn).toHaveBeenCalledWith(app)
    })
  })
})
