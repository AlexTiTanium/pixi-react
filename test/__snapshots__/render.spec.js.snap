// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render call injectDevtools 1`] = `
Array [
  Array [
    Object {
      "bundleType": 1,
      "findHostInstanceByFiber": [Function],
      "rendererPackageName": "@inlet/react-pixi",
      "version": "16.13.1",
    },
  ],
]
`;
